### ISSUE TITLE
Error while index creation error does not have any details attached

### ISSUE THREAD

**Issue by dobesv** [uken/fluent-plugin-elasticsearch#803]
#### Problem

Fluent fails to create the indexes but doesn't explain why.  I'm not really sure why the error doesn't print in this case, hopefully someone has a clue what kind of error doesn't show anything for `inspect`.

```
2020-09-09 00:10:37 +0000 [error]: [elasticsearch] Error while index creation - <logs.kube-system.kube-proxy.default.2020.09.08-000001>: #
2020-09-09 00:10:37 +0000 [warn]: [elasticsearch] failed to flush the buffer. retry_time=169 next_retry_seconds=2020-09-09 00:11:05 +0000 chunk="5aed593feb207ee373af7f8beab1da3c" error_class=Elasticsearch::Transport::Transport::Errors::NotFound error="[404] {\"error\":{\"root_cause\":[{\"type\":\"index_not_found_exception\",\"reason\":\"no such index [logs.kube-system.kube-proxy.default.2020.09.08-000001]\",\"index_uuid\":\"_na_\",\"resource.type\":\"index_or_alias\",\"resource.id\":\"logs.kube-system.kube-proxy.default.2020.09.08-000001\",\"index\":\"logs.kube-system.kube-proxy.default.2020.09.08-000001\"}],\"type\":\"index_not_found_exception\",\"reason\":\"no such index [logs.kube-system.kube-proxy.default.2020.09.08-000001]\",\"index_uuid\":\"_na_\",\"resource.type\":\"index_or_alias\",\"resource.id\":\"logs.kube-system.kube-proxy.default.2020.09.08-000001\",\"index\":\"logs.kube-system.kube-proxy.default.2020.09.08-000001\"},\"status\":404}"
```

#### Using Fluentd and ES plugin versions

* Kubernetes
* `fluentd-1.11.2 without supervision pid=1 ruby="2.6.6"`

Gemfile.lock:

```
GEM
  remote: https://rubygems.org/
  specs:
    activemodel (5.2.4.3)
      activesupport (= 5.2.4.3)
    activerecord (5.2.4.3)
      activemodel (= 5.2.4.3)
      activesupport (= 5.2.4.3)
      arel (>= 9.0)
    activerecord-import (0.28.2)
      activerecord (>= 3.2)
    activesupport (5.2.4.3)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (>= 0.7, < 2)
      minitest (~> 5.1)
      tzinfo (~> 1.1)
    addressable (2.7.0)
      public_suffix (>= 2.0.2, < 5.0)
    arel (9.0.0)
    async (1.24.2)
      console (~> 1.0)
      nio4r (~> 2.3)
      timers (~> 4.1)
    async-http (0.50.8)
      async (~> 1.23)
      async-io (~> 1.27.0)
      async-pool (~> 0.2)
      protocol-http (~> 0.15.1)
      protocol-http1 (~> 0.10.0)
      protocol-http2 (~> 0.11.0)
    async-io (1.27.7)
      async (~> 1.14)
    async-pool (0.2.0)
      async (~> 1.8)
    aws-eventstream (1.1.0)
    aws-partitions (1.363.0)
    aws-sdk-core (3.105.0)
      aws-eventstream (~> 1, >= 1.0.2)
      aws-partitions (~> 1, >= 1.239.0)
      aws-sigv4 (~> 1.1)
      jmespath (~> 1.0)
    aws-sdk-kms (1.37.0)
      aws-sdk-core (~> 3, >= 3.99.0)
      aws-sigv4 (~> 1.1)
    aws-sdk-s3 (1.79.1)
      aws-sdk-core (~> 3, >= 3.104.3)
      aws-sdk-kms (~> 1)
      aws-sigv4 (~> 1.1)
    aws-sdk-sqs (1.32.0)
      aws-sdk-core (~> 3, >= 3.99.0)
      aws-sigv4 (~> 1.1)
    aws-sigv4 (1.2.2)
      aws-eventstream (~> 1, >= 1.0.2)
    bigdecimal (2.0.0)
    concurrent-ruby (1.1.7)
    console (1.8.2)
    cool.io (1.6.0)
    declarative (0.0.20)
    declarative-option (0.1.0)
    domain_name (0.5.20190701)
      unf (>= 0.0.5, < 1.0.0)
    elasticsearch (7.9.0)
      elasticsearch-api (= 7.9.0)
      elasticsearch-transport (= 7.9.0)
    elasticsearch-api (7.9.0)
      multi_json
    elasticsearch-transport (7.9.0)
      faraday (~> 1)
      multi_json
    elasticsearch-xpack (7.9.0)
      elasticsearch-api (>= 6)
    excon (0.76.0)
    faraday (1.0.1)
      multipart-post (>= 1.2, < 3)
    ffi (1.13.1)
    ffi-compiler (1.0.1)
      ffi (>= 1.0.0)
      rake
    fluent-config-regexp-type (1.0.0)
      fluentd (> 1.0.0, < 2)
    fluent-plugin-bigquery (2.2.0)
      fluentd (>= 0.14.0, < 2)
      google-api-client (>= 0.11.0)
      googleauth (>= 0.5.0)
      multi_json
    fluent-plugin-concat (2.4.0)
      fluentd (>= 0.14.0, < 2)
    fluent-plugin-detect-exceptions (0.0.13)
      fluentd (>= 0.10)
    fluent-plugin-elasticsearch (4.1.2)
      elasticsearch
      excon
      fluentd (>= 0.14.22)
    fluent-plugin-grok-parser (2.6.1)
      fluentd (>= 0.14.6, < 2)
    fluent-plugin-json-in-json-2 (1.0.2)
      fluentd (>= 0.14.0, < 2)
      yajl-ruby (~> 1.0)
    fluent-plugin-kubernetes_metadata_filter (2.5.2)
      fluentd (>= 0.14.0, < 1.12)
      kubeclient (< 5)
      lru_redux
    fluent-plugin-multi-format-parser (1.0.0)
      fluentd (>= 0.14.0, < 2)
    fluent-plugin-mutate_filter (1.0.7)
      fluentd (>= 0.14.10, < 2)
    fluent-plugin-prometheus (1.8.3)
      fluentd (>= 1.9.1, < 2)
      prometheus-client (< 0.10)
    fluent-plugin-record-modifier (2.1.0)
      fluentd (>= 1.0, < 2)
    fluent-plugin-rewrite-tag-filter (2.3.0)
      fluent-config-regexp-type
      fluentd (>= 0.14.2, < 2)
    fluent-plugin-s3 (1.4.0)
      aws-sdk-s3 (~> 1.60)
      aws-sdk-sqs (~> 1.23)
      fluentd (>= 0.14.22, < 2)
    fluent-plugin-sql (2.0.0)
      activerecord (~> 5.1)
      activerecord-import (~> 0.7)
      fluentd (>= 0.12.17, < 2)
    fluent-plugin-systemd (1.0.2)
      fluentd (>= 0.14.11, < 2)
      systemd-journal (~> 1.3.2)
    fluentd (1.11.2)
      cool.io (>= 1.4.5, < 2.0.0)
      http_parser.rb (>= 0.5.1, < 0.7.0)
      msgpack (>= 1.3.1, < 2.0.0)
      serverengine (>= 2.0.4, < 3.0.0)
      sigdump (~> 0.2.2)
      strptime (>= 0.2.2, < 1.0.0)
      tzinfo (>= 1.0, < 3.0)
      tzinfo-data (~> 1.0)
      yajl-ruby (~> 1.0)
    google-api-client (0.44.0)
      addressable (~> 2.5, >= 2.5.1)
      googleauth (~> 0.9)
      httpclient (>= 2.8.1, < 3.0)
      mini_mime (~> 1.0)
      representable (~> 3.0)
      retriable (>= 2.0, < 4.0)
      signet (~> 0.12)
    googleauth (0.13.1)
      faraday (>= 0.17.3, < 2.0)
      jwt (>= 1.4, < 3.0)
      memoist (~> 0.16)
      multi_json (~> 1.11)
      os (>= 0.9, < 2.0)
      signet (~> 0.14)
    http (4.4.1)
      addressable (~> 2.3)
      http-cookie (~> 1.0)
      http-form_data (~> 2.2)
      http-parser (~> 1.2.0)
    http-accept (1.7.0)
    http-cookie (1.0.3)
      domain_name (~> 0.5)
    http-form_data (2.3.0)
    http-parser (1.2.1)
      ffi-compiler (>= 1.0, < 2.0)
    http_parser.rb (0.6.0)
    httpclient (2.8.3)
    i18n (1.8.5)
      concurrent-ruby (~> 1.0)
    jmespath (1.4.0)
    json (2.3.1)
    jsonpath (1.0.5)
      multi_json
      to_regexp (~> 0.2.1)
    jwt (2.2.2)
    kubeclient (4.9.1)
      http (>= 3.0, < 5.0)
      jsonpath (~> 1.0)
      recursive-open-struct (~> 1.1, >= 1.1.1)
      rest-client (~> 2.0)
    lru_redux (1.1.0)
    memoist (0.16.2)
    mime-types (3.3.1)
      mime-types-data (~> 3.2015)
    mime-types-data (3.2020.0512)
    mini_mime (1.0.2)
    minitest (5.14.2)
    msgpack (1.3.3)
    multi_json (1.15.0)
    multipart-post (2.1.1)
    netrc (0.11.0)
    nio4r (2.5.2)
    oj (3.10.13)
    os (1.1.1)
    pg (1.2.3)
    prometheus-client (0.9.0)
      quantile (~> 0.2.1)
    protocol-hpack (1.4.2)
    protocol-http (0.15.1)
    protocol-http1 (0.10.3)
      protocol-http (~> 0.15)
    protocol-http2 (0.11.6)
      protocol-hpack (~> 1.4)
      protocol-http (~> 0.15)
    public_suffix (4.0.6)
    quantile (0.2.1)
    rake (13.0.1)
    recursive-open-struct (1.1.2)
    representable (3.0.4)
      declarative (< 0.1.0)
      declarative-option (< 0.2.0)
      uber (< 0.2.0)
    rest-client (2.1.0)
      http-accept (>= 1.7.0, < 2.0)
      http-cookie (>= 1.0.2, < 2.0)
      mime-types (>= 1.16, < 4.0)
      netrc (~> 0.8)
    retriable (3.1.2)
    serverengine (2.2.1)
      sigdump (~> 0.2.2)
    sigdump (0.2.4)
    signet (0.14.0)
      addressable (~> 2.3)
      faraday (>= 0.17.3, < 2.0)
      jwt (>= 1.5, < 3.0)
      multi_json (~> 1.10)
    strptime (0.2.4)
    systemd-journal (1.3.3)
      ffi (~> 1.9)
    thread_safe (0.3.6)
    timers (4.3.0)
    to_regexp (0.2.1)
    tzinfo (1.2.7)
      thread_safe (~> 0.1)
    tzinfo-data (1.2020.1)
      tzinfo (>= 1.0.0)
    uber (0.1.0)
    unf (0.1.4)
      unf_ext
    unf_ext (0.0.7.7)
    yajl-ruby (1.4.1)

PLATFORMS
  ruby

DEPENDENCIES
  async-http
  bigdecimal
  elasticsearch
  elasticsearch-xpack
  ffi
  fluent-plugin-bigquery
  fluent-plugin-concat
  fluent-plugin-detect-exceptions
  fluent-plugin-elasticsearch
  fluent-plugin-grok-parser
  fluent-plugin-json-in-json-2
  fluent-plugin-kubernetes_metadata_filter
  fluent-plugin-multi-format-parser
  fluent-plugin-mutate_filter
  fluent-plugin-prometheus
  fluent-plugin-record-modifier
  fluent-plugin-rewrite-tag-filter
  fluent-plugin-s3
  fluent-plugin-sql
  fluent-plugin-systemd
  fluentd
  json
  oj
  pg

BUNDLED WITH
   2.0.2
```

* ES version 7.8.0

Config block:

```
<label @elasticsearch>
  <match **>
    @id elasticsearch
    @type elasticsearch
    @log_level info
    include_tag_key true
    host elasticsearch.elasticsearch
    port 9200
    default_elasticsearch_version 7
    verify_es_version_at_startup true
    suppress_type_name true
    logstash_format true
    logstash_prefix logs.${tag}
    logstash_prefix_separator .
    enable_ilm true
    ilm_policy_id fluentd-logs
    ilm_policy {"policy":{"phases":{"hot":{"min_age":"0ms","actions":{"rollover":{"max_age":"1d","max_size":"20gb"},"set_priority":{"priority":100}}},"warm":{"actions":{"set_priority":{"priority":50}}},"delete":{"min_age":"14d","actions":{"delete":{}}}}}}
    ilm_policy_overwrite false
    log_es_400_reason true
    templates { "logs": "/etc/fluent/config.d/logs-es-template.json", "formative-logs": "/etc/fluent/config.d/formative-es-template.json", "webclient-logs": "/etc/fluent/config.d/webclient-es-template.json", "haproxy-access-logs": "/etc/fluent/config.d/logs-haproxy-access-es-template.json" }
    template_overwrite true
    template_name logs
    template_file "/etc/fluent/config.d/logs-es-template.json"

    <buffer tag>
      @type file
      path /var/log/fluentd-buffers/kubernetes.${tag}
      flush_mode interval
      retry_max_interval 30
      flush_thread_count 2
    </buffer>
  </match>
</label>
```

Template:

```
{
  "index_patterns": ["logs.*"],
  "mappings": {
    "date_detection": false,
    "dynamic_templates": [
      {
        "strings_as_keyword": {
          "mapping": { "ignore_above": 1024, "type": "keyword" },
          "match_mapping_type": "string"
        }
      },
      {
        "count_suffix_as_long": {
          "mapping": { "type": "long" },
          "match": "*Count",
          "match_mapping_type": "string"
        }
      },
      {
        "disable_unmapped_objects": {
          "mapping": { "enabled": false },
          "match_mapping_type": "object"
        }
      }
    ],
    "properties": {
      "@timestamp": { "type": "date" },
      "err": {
        "properties": {
          "code": { "ignore_above": 1024, "type": "keyword" },
          "id": { "ignore_above": 1024, "type": "keyword" },
          "message": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword"
              }
            }
          },
          "msg": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword"
              }
            }
          },
          "stack": { "norms": false, "type": "text" }
        }
      },
      "message": {
        "norms": false,
        "type": "text",
        "fields": {
          "keyword": {
            "type": "keyword"
          }
        }
      },
      "stream": { "ignore_above": 1024, "type": "keyword" },
      "docker": { "properties": { "container_id": { "ignore_above": 1024, "type": "keyword" } } },
      "pid": { "type": "long" },
      "port": { "type": "long" },
      "level": { "type": "long" },
      "levelName": { "type": "keyword" },
      "msg": {
        "type": "text",
        "fields": {
          "keyword": {
            "type": "keyword"
          }
        }
      },
      "log": {
        "type": "text",
        "fields": {
          "keyword": {
            "type": "keyword"
          }
        }
      },
      "kubernetes": {
        "properties": {
          "container_image": { "ignore_above": 1024, "type": "keyword" },
          "container_image_id": { "ignore_above": 1024, "type": "keyword" },
          "container_name": { "ignore_above": 1024, "type": "keyword" },
          "host": { "ignore_above": 1024, "type": "keyword" },
          "master_url": { "ignore_above": 1024, "type": "keyword" },
          "namespace_id": { "ignore_above": 1024, "type": "keyword" },
          "namespace_name": { "ignore_above": 1024, "type": "keyword" },
          "pod_name": { "ignore_above": 1024, "type": "keyword" },
          "pod_id": { "ignore_above": 1024, "type": "keyword" },
          "labels": {
            "properties": {
              "app": { "ignore_above": 1024, "type": "keyword" },
              "k8s-app": { "ignore_above": 1024, "type": "keyword" },
              "formative-image": { "ignore_above": 1024, "type": "keyword" },
              "pod-template-hash": { "ignore_above": 1024, "type": "keyword" }
            }
          }
        }
      }
    }
  },
  "order": 1,
  "settings": {
    "index": {
      "mapping": { "total_fields": { "limit": 10000 } },
      "number_of_shards": 3,
      "refresh_interval": "5s"
    }
  }
}

```

**Comment by dobesv** [comment]
After some investigation, the error that is not being printed is that the cluster has reached its maximum shards per node.

I can increase the maximum shards per node or run more nodes, but it would be nice if the error were printed properly in the logs.

This might mean that you could reproduce this by running an ES cluster with max_shards_per_node set low (like 20) and set number_of_shards in the index template high (like 20).

**Comment by cosmo0920** [comment]
We do not provide data manipulation services nor operate very large ES cluster....
I'm using this plugin on minikube k8s and a few Elasticsearch nodes (2~4 nodes).
Our main target is fluentd-kubernetes-daemonset data ingestion with fluentd-debian-elasticsearch image....

The following error should be provided from Elasticsearch?

> ```
> 2020-09-09 00:10:37 +0000 [error]: [elasticsearch] Error while index creation - <logs.kube-system.kube-proxy.default.2020.09.08-000001>: #
> 2020-09-09 00:10:37 +0000 [warn]: [elasticsearch] failed to flush the buffer. retry_time=169 next_retry_seconds=2020-09-09 00:11:05 +0000 chunk="5aed593feb207ee373af7f8beab1da3c" error_class=Elasticsearch::Transport::Transport::Errors::NotFound error="[404] {\"error\":{\"root_cause\":[{\"type\":\"index_not_found_exception\",\"reason\":\"no such index [logs.kube-system.kube-proxy.default.2020.09.08-000001]\",\"index_uuid\":\"_na_\",\"resource.type\":\"index_or_alias\",\"resource.id\":\"logs.kube-system.kube-proxy.default.2020.09.08-000001\",\"index\":\"logs.kube-system.kube-proxy.default.2020.09.08-000001\"}],\"type\":\"index_not_found_exception\",\"reason\":\"no such index [logs.kube-system.kube-proxy.default.2020.09.08-000001]\",\"index_uuid\":\"_na_\",\"resource.type\":\"index_or_alias\",\"resource.id\":\"logs.kube-system.kube-proxy.default.2020.09.08-000001\",\"index\":\"logs.kube-system.kube-proxy.default.2020.09.08-000001\"},\"status\":404}"
```

If so, ES plugin does not store Elasticsearch state. So, ES plugin just complains with messages which is from Elasticsearch....
I don't think that ES plugin can complain properly in this situation.

**Comment by dobesv** [comment]
The errors I put in there are logged by fluentd here:

https://github.com/uken/fluent-plugin-elasticsearch/blob/9976147cb422caa8b58d7a6118146dd00f2e44f8/lib/fluent/plugin/elasticsearch_index_template.rb#L64

what I don't understand is why it just prints `#` where it should have printed the actual error.  Does this mean elasticsearch returned an empty object or empty string instead of an error?

**Comment by dobesv** [comment]
I wonder if something like `router.emit_error_event(@time_parse_error_tag, Fluent::Engine.now, {'index_name' => index_name, 'message' => e.message}, e)` or `log.error("Error creating index #{index_name}: #{e.message}")` would work?

**Comment by cosmo0920** [comment]
`log.error("Error while index creation - #{index_name}.", error: e)` would be better in this case:
log#error implementation: https://github.com/fluent/fluentd/blob/master/lib/fluent/log.rb#L369

**Comment by cosmo0920** [comment]
I'd released a patch for fixing this issue as v4.1.4.
